File: nvpl_blas/c/dtpsv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dtpsv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transa = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_DIAG diag = CblasNonUnit;
    double * AP = NULL;
    double * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_ap = (N * (N + 1)) / 2;

    printf("\nExample: cblas_dtpsv for solving a system of linear equations whose coefficients are"
            " in a triangular packed matrix.\n\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", order=%c"
            ", uplo=%c, transa=%c, diag=%c\n", (int64_t)N, (int64_t)incX,
            order_to_char(order), uplo_to_char(uplo), transpose_to_char(transa), diag_to_char(diag));

    // allocate memory
    AP = (double *)malloc(len_ap  * sizeof(double));
    X = (double *)malloc(len_x * sizeof(double));

    // fill data
    fill_dvector(AP, len_ap, 1);
    fill_dvector(X, N, incX);

    // print input data
    print_dvector(AP, len_ap, 1, "AP");
    print_dvector(X, N, incX, "X");

    // call cblas_dtpsv
    cblas_dtpsv(order, uplo, transa, diag, N, AP, X, incX);

    // print result
    printf("\nAfter dtpsv operation:\n");
    print_dvector(X, N, incX, "X");

    // release memory
    free(AP);
    free(X);
    return 0;
}

File: nvpl_blas/f77/dtpsv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DTPSV 
!
!******************************************************************************
      program   DTPSV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1     uplo, transa, diag
      parameter       (uplo='U', transa='N', diag='N')
      integer         n, incx, i, j
      parameter       (n=2, incx=1)
      integer         xsize, apsize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 apsize=(n*(n+1))/2)
      double precision x(xsize), ap(apsize)

*       External Subroutines
      external        print_dvector, fill_dvector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, uplo, transa, diag
      call fill_dvector(x,n,incx)
      call fill_dvector(ap,apsize,1)
      call print_dvector(ap,apsize,1,'AP')
      call print_dvector(x,n,incx,'X')
*
*      Call DTPSV subroutine
      call DTPSV(uplo,transa,diag,n,ap,x,incx)
*
      print*
      print 102, 'After DTPSV operation:'
      call print_dvector(x,n,incx,'X')

      stop
 99   format('Example: DTPSV for solving a system of linear equations'
     &       ' whose coefficients are in a triangular packed matrix.')
 100  format('#### args: n=',i1,', incx=',i1,
     &       ', uplo=',a1,', transa=',a1,', diag=',a1)
 102  format(A)
      end
